src/main/java/org/candlepin/model/RhsmApiCompatCurator.java

```diff
@@ -103,4 +116,421 @@ public List<ConsumerEntitlementCount> getConsumerEntitlementCounts(Collection<St
+                "FROM cp_consumer hypervisor " +
+                "JOIN cp_consumer_guests guest ON guest.consumer_id = hypervisor.id " +
+                "JOIN ( " +
+                "    SELECT guest_id, MAX(updated) AS last_updated " +
+                "    FROM cp_consumer_guests " +
+                "    GROUP BY guest_id " +
+                ") glu " +
+                "ON glu.guest_id = guest.guest_id AND glu.last_updated = guest.updated " +
+                "WHERE hypervisor.owner_id = :owner_id";
+        }
+
+        List<Object[]> rows = entityManager.createNativeQuery(sql)
+            .setParameter("owner_id", ownerId)
+            .getResultList();
+
+        Map<String, HypervisorGuestMapping> result = new HashMap<>();
+        for (Object[] row : rows) {
+            String hypervisorId = (String) row[0];
+            String hypervisorUuid = (String) row[1];
+            String hypervisorName = (String) row[2];
+            String guestId = (String) row[3];
+            HypervisorGuestMapping mapping = new HypervisorGuestMapping(
+                hypervisorId, hypervisorName, hypervisorUuid, guestId);
+
+            // Since we know there will be few consumer records, we can preload all mappings.
+            // This avoids looping through all consumer feeds for each lookup and improves performance.
+            List<String> guestIdVariants = Util.getPossibleUuids(guestId);
+            for (String variant : guestIdVariants) {
+                result.put(variant.toLowerCase(), mapping);
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Retrieves a mapping from consumer ID to a list of {@link ConsumerInstalledProduct} for all
+     * installed products belonging to consumers matching the specified owner and filter criteria.
+     *
+     * @param ownerId      the ID of the owner whose consumers' products are to be fetched; must not be null
+     * @param afterId      (optional) only consumers with an ID lexicographically greater than this value
+     *                     will be included
+     * @param afterUuid    (optional) only consumers with a UUID lexicographically greater than this value
+     *                     will be included
+     * @param afterCheckin (optional) only consumers with a last check-in timestamp after this value will
+     *                     be included
+     * @param page         the page number to retrieve
+     * @param perPage      the number of consumers to return per page
+     * @return a map where each key is a consumer ID and each value is a list of installed products
+     * for that consumer
+     */
+    private Map<String, Set<ConsumerFeedInstalledProduct>> getInstalledProductsByConsumer(String ownerId,
+        String afterId, String afterUuid, OffsetDateTime afterCheckin, int page, int perPage) {
+        String cte = buildConsumersSubquery(afterId, afterUuid, afterCheckin);
```

Comment:
We should update the variable naming here to better indicate what the value is. Even if the subquery were a CTE, that's not important in this context.